### ISSUE TITLE
optimizer: JoinImplementation transform needs several iterations to settle on a join plan

### ISSUE THREAD

**Issue by asenac** [MaterializeInc/materialize#7501]
As shown in the snippet below, the `ArrangeBy` operators added by `JoinImplementation` may affect the plan decision taken by a later invocation of `JoinImplementation`.

```
%0 =
| Get ? (u1)

%1 =
| Get ? (u3)

%2 =
| Get ? (u7)
| Filter !(isnull(#0)), (#7 <= "o"), (#7 >= "g")

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Unimplemented
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
JoinImplementation (lifts the predicate because it needs to add an arrangement)
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| Filter !(isnull(#25)), (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
ColumnKnowledge
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| Filter true, (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
FoldConstants
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| Filter (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
Demand
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| | demand = (#0..#24, #26..#29, #31, #32)
| Filter (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
JoinImplementation
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %2 %0.(#5) %1.(#1)
| | demand = (#0..#24, #26..#29, #31, #32)
| Filter (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)
```

_Originally posted by @asenac in https://github.com/MaterializeInc/materialize/issues/7470#issuecomment-884971201_

**Comment by asenac** [comment]
Scenario in [#7470 (comment)](https://github.com/MaterializeInc/materialize/pull/7470#issuecomment-884797331)

**Comment by wangandi** [comment]
I found a case where running `JoinImplementation` in a loop may theoretically result in an improved plan. 

In https://github.com/MaterializeInc/materialize/pull/7864#issuecomment-900936867, there is an inner join and an outer join, The inner join refers to `Project{Get(t1), []}` whereas the outer join refers to `Get(t1)` without a project around it.

If
* we ran `JoinImplementation` in a loop
* `JoinImplementation` supplied information on all arrangements in the `MirRelationExpr` (it does not do that right now).

Then during the second iteration, the inner join could see that an `{ArrangeBy{Get{t1}, [[]]}` is being created for the outer join and decide to use the same arrangement.

But in this case, I still think a better alternative would be for JoinImplementation to look at all joins in the `MirRelationExpr` at once to see whether arrangements could be between shared between joins.
